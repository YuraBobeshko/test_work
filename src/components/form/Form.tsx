import React from "react";
import { useForm } from "react-hook-form";

import Input from "../Input/Input";
import "./form.scss";

const Form: React.FC = () => {
  const { handleSubmit, register, errors } = useForm({});
  const onSubmit = (values: any) => {
    console.info(values);
  };

  return (
    <div className={"wrapper-form"}>
      <p className={"form__heading"}>Log in to Example</p>
      <a className={"form__link-head"} href="/">
        or create an account
      </a>
      <form className={"form"} onSubmit={handleSubmit(onSubmit)}>
        <Input
          name={"e.g.john.doe@gmail.com"}
          type={"text"}
          label={"E-mail"}
          errors={errors}
          register={register({
            required: "Required",
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
              message: "error email"
            }
          })}
        />
        <Input
          name={"●●●●●●●●●●●●●"}
          type={"password"}
          label={"Password"}
          errors={errors}
          register={register({
            required: "Required",
            minLength: {
              value: 8,
              message: "error password"
            },
            max: {
              value: 20,
              message: "error password"
            }
          })}
        />
        <Input
          name={"checkbox"}
          type={"checkbox"}
          customLabel={<p>Agree with <a href={'/'}>Terms & Conditions</a></p>}
          errors={errors}
          register={register({
            required: "is required"
          })}
        />
        <button className={"form-btn"} type="submit">
          Login
        </button>
      </form>
      <a className={"form__link"} href="/" onClick={() => alert("it's not my problem")}>Forgotten password?</a>
    </div>
  );
};

export default Form;
