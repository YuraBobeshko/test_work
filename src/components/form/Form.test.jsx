import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { cleanup } from "@testing-library/react";

import Form from "./Form";

afterEach(cleanup);

const wrapper = shallow(<Form />);
const wrapperForm = wrapper.find("form").children();
const wrapperInput = wrapper
  .find("form")
  .children()
  .find("Input")
  .children();
const FormComponent = renderer.create(<Form />).toJSON();
const FormComponentError = renderer.create(<Form />).toJSON();

describe("should structure <Form /> be correct", () => {
  test("is form", () => {
    expect(wrapper.find("form")).toHaveLength(1);
  });

  test("is button", () => {
    expect(wrapperForm.find("button")).toHaveLength(1);
  });

  test("all 3 <Input />", () => {
    expect(wrapperForm.find("Input")).toHaveLength(3);
  });

  test("all 3 input", () => {
    wrapperInput.find("Input").forEach(node => {
      expect(
        node
          .find("div")
          .children()
          .find("input")
      ).toHaveLength(1);
    });
  });
});

describe("should luck correct", () => {
  expect(FormComponent).toMatchSnapshot();
});

describe("Form changes the text after click", () => {
  test("is form with error", () => {
    wrapperForm.find("button").simulate("click");
    wrapperInput.find("Input").forEach(node => {
      expect(
        node
          .find("div")
          .children()
          .find("span.error")
      ).toHaveLength(1);
    });
    expect(FormComponentError).toMatchSnapshot();
  });
});
