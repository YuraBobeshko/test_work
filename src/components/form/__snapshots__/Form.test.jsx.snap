// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__heading"
  >
    Log in to Example
  </p>
  <a
    className="form__link-head"
    href="/"
  >
    or create an account
  </a>
  <div
    className="content"
  >
    <input
      className=""
      id="email"
      name="email"
      placeholder="e.g.john.doe@gmail.com"
      type="text"
    />
    <label
      className="label"
      htmlFor="email"
    >
      E-mail
    </label>
  </div>
  <div
    className="content"
  >
    <input
      className=""
      id="password"
      name="password"
      placeholder="●●●●●●●●●●●●●"
      type="password"
    />
    <label
      className="label"
      htmlFor="password"
    >
      Password
    </label>
  </div>
  <div
    className="content-checkbox"
  >
    <input
      className=""
      id="agree"
      name="agree"
      type="checkbox"
    />
    <label
      className="label"
      htmlFor="agree"
    >
      <p>
        Agree with 
        <a
          href="/"
        >
          Terms & Conditions
        </a>
      </p>
    </label>
  </div>
  <button
    className="btn form-btn"
    type="submit"
  >
    Login
  </button>
  <a
    className="form__link"
    href="/"
    onClick={[Function]}
  >
    Forgotten password?
  </a>
</form>
`;

exports[`Form changes the text after click is form with error 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__heading"
  >
    Log in to Example
  </p>
  <a
    className="form__link-head"
    href="/"
  >
    or create an account
  </a>
  <div
    className="content"
  >
    <input
      className=""
      id="email"
      name="email"
      placeholder="e.g.john.doe@gmail.com"
      type="text"
    />
    <label
      className="label"
      htmlFor="email"
    >
      E-mail
    </label>
  </div>
  <div
    className="content"
  >
    <input
      className=""
      id="password"
      name="password"
      placeholder="●●●●●●●●●●●●●"
      type="password"
    />
    <label
      className="label"
      htmlFor="password"
    >
      Password
    </label>
  </div>
  <div
    className="content-checkbox"
  >
    <input
      className=""
      id="agree"
      name="agree"
      type="checkbox"
    />
    <label
      className="label"
      htmlFor="agree"
    >
      <p>
        Agree with 
        <a
          href="/"
        >
          Terms & Conditions
        </a>
      </p>
    </label>
  </div>
  <button
    className="btn form-btn"
    type="submit"
  >
    Login
  </button>
  <a
    className="form__link"
    href="/"
    onClick={[Function]}
  >
    Forgotten password?
  </a>
</form>
`;
